bacula/manpages/bregex.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH BREGEX 8 "30 October 2011" "Kern Sibbald" "Network backup, utilities"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
 bregex \- Bacula's 'regex' engine
.SH SYNOPSIS
.B bregex
.RI [ options ]
.I -f <data\-file>
.br
.SH DESCRIPTION
This manual page documents briefly the
.B bregex
command.
.br
This program can be useful for testing regex expressions to be applied against a list of filenames.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invoke bold face and italics,
.\" respectively.
.SH OPTIONS
A summary of options is included below.
.TP
.B \-?
Show version and usage of program.
.TP
.BI \-d\  nn
Set debug level to \fInn\fP.
.TP
.BI \-dt
Print timestamp in debug output
.TP
.BI \-f\ <data\-file>
The data-file is a filename that contains lines of data to be matched (or not) against one or more patterns. When the program is run, it will prompt you for a regular expression pattern, then apply it one line at a time against the data in the file. Each line that matches will be printed preceded by its line number. You will then be prompted again for another pattern.
.TP
.BI \-n
Print lines that do not match
.TP
.BI \-l
Suppress lines numbers.
.SH SEE ALSO
.BR regex(7)
.br
.SH AUTHOR
This manual page was written by Bruno Friedmann
.nh
<bruno@ioda\-net.ch>.


bacula/ChangeLog
              Changelog on version 7.0.5

Release version 7.0.5

28Jul14
 - Fix #547 by adding .schedule command

27Jul14
 - Update AUTHORS
 - Fix bug #2079 with patch from Robert Oschwald

26Jul14
 - Fix orphaned file descriptors during errors
 - Yet another client==NULL
 - Improve FD and SD cancel
 - Jim Raney's TLS patch
 - Update AUTHORS
 - Fix bug #1679 pool overrides not shown in manual run display
 - Attempt to avoid client==NULL

23Jul14
 - Fix for bug #2082 (hopefully)

22Jul14
 - Fix seg fault in jobq.c

14Jul14
 - make stop after first error

12Jul14
 - Increase status schedule days from 500 to 3000
 - Remove bad cherry-pick
 - Fix compiler warning
 - Allow options create_postgresql_database from patch in bug #2075 by roos
 - Fix bug #2074 crashes when no conf file present

10Jul14
 - Set pthread id in jcr at beginning so the job can be canceled.
 - Fix possible heartbeat interval timing problems

08Jul14
 - Fix some errors reported by valgrind. May fix the problem with bsmtp command.
 - Ensure b_sterror() is using the correct size of input buffer

07Jul14
 - Fix possible seg fault

04Jul14
 - Fix segfault when trying to stop the bnet_server thread in terminate_stored()

03Jul14
 - Fix bad link bug #2076

02Jul14
 - Fix compilation of bsock.c when TLS is not available

27Jun14
 - Correct L suffix to be LL
 - Fix bad copy/migrate data header

26Jun14
 - On termination shutdown thread server

23Jun14
 - baculum: Updated README file
 - baculum: Update English language texts
 - baculum: Saving auth file for web server HTTP Basic auth
 - baculum: Added directory for web server logs
 - baculum: Added example Lighttpd configuration for Baculum and sample web 
  server auth file
 - Expanded auth error message
 - baculum: Support for web servers which do not provide direct info about HTTP 
  Basic auth

15Jun14
 - Fix limit bandwidth calculation
 - Eliminate strcpy() from bsmtp

12Jun14
 - Fix for configuring sudo option for bconsole access
 - Display correct NextPool overrides + use Job NextPool in restore if available

09Jun14
 - Fix Bacula to work with newer Windows pthreads library

17May14
 - Fix bug #180 ERR=success in btape when tape error

Bugs fixed/closed since last release:
1679 180 2074 2075 2076 2079 2082 547

====

Release version 7.0.4
02Jun14
 - Better error handling for cancel command

01Jun14
 - Fix compiler warning + simplify some #ifdefs

22May14
 - Fix copy/migration to second SD

19May14
 - Fix calls to sl.set_string()
 - Improve sellist code

=====

Release version 7.0.3
12May14
 - Fix error handling in do_alist_prompt
 - Tighten error condition handling in sellist
 - Add new cancel test

06May14
 - Update LICENSE and LICENSE-FAQ

03May14
 - Also update autoconf/aclocal.m4
 - Reschedule on error caused EndTime to be incorrect -- fixes bug #2029
 - Flush console queued job messages -- should fix bug #2054
 - Attempt to fix FreeBSD echo/printf, bug #2048
 - Update to newer libtool + config.guess
 - Recompile configure
 - Apply fix supplied for acl.c in bug #2050

01May14
 - Fix a SD seg fault that occurs with over committed drives

28Apr14
 - Clear bvfs cache and set debug options available only for admin
 - Moved auth params to curl opts
 - Filtred single results for restricted consoles

27Apr14
 - Removed unnecessary debug
 - Changed e-mail address in gettext file
 - Support for customized and restricted consoles

15Apr14
 - Misc changes for rpm building (made by Louis)

13Apr14
 - Updated requirements for Baculum

12Apr14
 - Apply fix for bug 2049: wrong drive selected

11Apr14
 - Fix #2047 about bthread_cond_wait_p not declared

09Apr14
 - Fix Bacula bug #2044 -- fix Makefile for bplugininfo linking
 - Fix Bacula bug #2046 -- sellist limited to 10000
 - Fix Bacula bug #2045 -- multiply defined daemon_event
 - Fix Bacula bug #2020 overflow in btape -- Andreas Koch

Bugs fixed/closed since last release:
2020 2029 2044 2045 2046 2047 2048 2050 2054


Release version 7.0.2
02Apr14
 - Remove more vestiges of libbacpy
 - Put back @PYTHON@ path in configure
 - Fix improper string in parser
 - Remove libbacpy from rpm spec files
 - Fix linking check_bacula
 - Fix new SD login in check_bacula
 - Tweek docs build process

Release version 7.0.1
31Mar14
 - Remove old plugin-test
 - Update po files
 - Enable installation of the bpluginfo utility
 - More tray-monitor updates
 - Add Simone Caronii to AUTHORS
 - Align command line switches in manpages.
 - Apply upgrade to config.guess
 - Remove bgnome-console and bwx-console leftovers.
 - Update tray-monitor header also for new bsock calls
 - Attempt to fix nagios to use new bsock calls
 - Update tray-monitor to new bsock calls

==========
Release version 7.0.0

24Mar14
 - Add Josip Almasi to AUTHORS
 - [PATCH] Support for restricted consoles in BAT config
 - [PATCH] Fix for free director directive
 - [PATCH] Fix auto-select restricted console for director in bconsole
 - Realign output display
 - Update ua_output.c from Branch-6.7
 - Add some missing Branch-6.7 updates
 - Added needed empty directories to Baculum
 - Fix for support PostgreSQL, MySQL and SQLite
 - Framework adjusting to Baculum database connections
 - Framework fix for lower case tables names in MySQL
 - Fix for Baculum SQLite support
 - Initial commit Baculum
 - Add Marcin to AUTHORS file
 - Strip trailing blanks
 - Update copyright year
 - Update LICENSE and header files
 - Remove old file
 - Add new header in misc files
 - Remove tray-monitor bwx-console manual installation
 - Remove FD python and examples
 - Fixup spec files
 - Remove pythonlib from lib
 - Update package-list
 - Fix SDCallsClient daemon synchronization
 - Add debug code + make 127.0.0.1 same as localhost for tls tests
 - Fix multiple DIRs in console
 - Make failure for bat to connect to DIR non-fatal
 - Fix bat style to one that works
 - Take disk-changer from Branch-6.7
 - Simplify Version output
 - Fix FDVersion for SD Calls Client test
 - Update accurate test
 - Update differential test

bacula/src/tools/bregex.c
/*
 * Test program for testing regular expressions.
 *
 *  Kern Sibbald, MMVI
 *
 */
/*
   Bacula® - The Network Backup Solution

   Copyright (C) 2006-2014 Free Software Foundation Europe e.V.

   The main author of Bacula is Kern Sibbald, with contributions from many
   others, a complete list can be found in the file AUTHORS.

   You may use this file and others of this release according to the
   license defined in the LICENSE file, which includes the Affero General
   Public License, v3.0 ("AGPLv3") and some additional permissions and
   terms pursuant to its AGPLv3 Section 7.

   Bacula® is a registered trademark of Kern Sibbald.
*/

#include "bacula.h"

/*
 *  If you define BACULA_REGEX, bregex will be built with the
 *  Bacula bregex library, which is the same code that we
 *  use on Win32, thus using Linux, you can test your Win32
 *  expressions. Otherwise, this program will link with the
 *  system library routines.
 */
//#define BACULA_REGEX

#ifdef BACULA_REGEX

#include "lib/bregex.h"

#else
#ifndef HAVE_REGEX_H
#include "lib/bregex.h"
#else
#include <regex.h>
#endif

#endif


static void usage()
{
   fprintf(stderr,
"\n"
"Usage: bregex [-d debug_level] -f <data-file>\n"
"       -f          specify file of data to be matched\n"
"       -l          suppress line numbers\n"
"       -n          print lines that do not match\n"
"       -d <nn>     set debug level to <nn>\n"
"       -dt         print timestamp in debug output\n"
"       -?          print this message.\n"
"\n\n");

   exit(1);
}


int main(int argc, char *const *argv)
{
   regex_t preg;
   char prbuf[500];
   char *fname = NULL;
   int rc, ch;
   char data[1000];
   char pat[500];
   FILE *fd;
   bool match_only = true;
   int lineno;
   bool no_linenos = false;


   setlocale(LC_ALL, "");
   bindtextdomain("bacula", LOCALEDIR);
   textdomain("bacula");

   while ((ch = getopt(argc, argv, "d:f:n?")) != -1) {
      switch (ch) {
      case 'd':                       /* set debug level */
         if (*optarg == 't') {
            dbg_timestamp = true;
         } else {
            debug_level = atoi(optarg);
            if (debug_level <= 0) {
               debug_level = 1;
            }
         }
         break;

      case 'f':                       /* data */
         fname = optarg;
         break;

      case 'l':
         no_linenos = true;
         break;

      case 'n':
         match_only = false;
         break;

      case '?':
      default:
         usage();

      }
   }
   argc -= optind;
   argv += optind;

   if (!fname) {
      printf("A data file must be specified.\n");
      usage();
   }

   OSDependentInit();

   for ( ;; ) {
      printf("Enter regex pattern: ");
      if (fgets(pat, sizeof(pat)-1, stdin) == NULL) {
         break;
      }
      strip_trailing_newline(pat);
      if (pat[0] == 0) {
         exit(0);
      }
      rc = regcomp(&preg, pat, REG_EXTENDED);
      if (rc != 0) {
         regerror(rc, &preg, prbuf, sizeof(prbuf));
         printf("Regex compile error: %s\n", prbuf);
         continue;
      }
      fd = fopen(fname, "r");
      if (!fd) {
         printf(_("Could not open data file: %s\n"), fname);
         exit(1);
      }
      lineno = 0;
      while (fgets(data, sizeof(data)-1, fd)) {
         const int nmatch = 30;
         regmatch_t pmatch[nmatch];
         strip_trailing_newline(data);
         lineno++;
         rc = regexec(&preg, data, nmatch, pmatch,  0);
         if ((match_only && rc == 0) || (!match_only && rc != 0)) {
            if (no_linenos) {
               printf("%s\n", data);
            } else {
               printf("%5d: %s\n", lineno, data);
            }
         }
      }
      fclose(fd);
      regfree(&preg);
   }
   exit(0);
}
